pkg/terraform/shared_grpc_server.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2022 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package terraform
+
+import (
+	"bufio"
+	"os"
+	"regexp"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	envReattachConfig = "TF_REATTACH_PROVIDERS"
+	regexReattachLine = envReattachConfig + `='(.*)'`
+	reattachTimeout   = 1 * time.Minute
+)
+
+func (ws *WorkspaceStore) startSharedServer() error { //nolint:gocyclo
+	if len(ws.nativeProviderPath) == 0 {
+		return nil
+	}
+	log := ws.logger.WithValues("nativeProviderPath", ws.nativeProviderPath, "nativeProviderArgs", ws.nativeProviderArgs)
+	if ws.reattachConfig != "" {
+		log.Debug("Shared gRPC server is running...", "reattachConfig", ws.reattachConfig)
+		return nil
+	}
+	errCh := make(chan error, 1)
+	reattachCh := make(chan string, 1)
+	re, err := regexp.Compile(regexReattachLine)
+	if err != nil {
+		return errors.Wrap(err, "failed to compile regexp")
+	}
+
+	go func() {
+		defer close(errCh)
+		defer close(reattachCh)
+		defer func() {
+			ws.mu.Lock()
+			ws.reattachConfig = ""
+			ws.mu.Unlock()
+		}()
+		//#nosec G204 no user input
+		cmd := ws.executor.Command(ws.nativeProviderPath, ws.nativeProviderArgs...)
```

Comment:
A reminder for my comment [here](https://github.com/crossplane/terrajet/issues/261#issuecomment-1070827997). We should remove env var-based auth slowly since it won't work with this setup.

Reply:
Actually I was meaning deprecating [this field](https://github.com/crossplane/terrajet/blob/1dfbdfcd7162105a67bd7ffb53b689f46a150450/pkg/terraform/store.go#L56) pointing that the configuration should be used instead of env.

Reply:
Oh. I see. I was considering `Setup.Env` as a general purpose configuration field that would allow us to inject arbitrary env. variables, not just those used for injecting authn env. variables. What do you think?

Reply:
I'm not sure. It's only used for credentials and they have to go to provider, so there isn't a use case coming to my mind that you want to affect TF CLI's env vars.